src/lib.rs

```diff
@@ -1410,6 +1544,348 @@ where
+                }
+                Err(DeviceCodeAction::IncreaseIntervalThenRetry) => {
+                    // Increase the interval, then wait that long.
+                    interval = interval + Duration::from_secs(5);
+                }
+                Err(DeviceCodeAction::DoubleIntervalThenRetry) => {
+                    // Double the interval, then wait that long.
+                    interval = interval.checked_mul(2).ok_or(RequestTokenError::Other(
+                        "Failed to increase interval".to_string(),
+                    ))?;
+                }
+                Err(DeviceCodeAction::NoFurtherRequests(req)) => break token_response(req),
+            };
+
+            // Actually sleep here.
+            std::thread::sleep(interval);
+            elapsed = elapsed + interval;
+        }
+    }
+
+    ///
+    /// Asynchronously sends the request to the authorization server and awaits a response.
+    ///
+    pub async fn request_async<C, F, RE>(
+        self,
+        http_client: C,
+    ) -> Result<TR, RequestTokenError<RE, TE>>
+    where
+        C: Fn(HttpRequest) -> F,
+        F: Future<Output = Result<HttpResponse, RE>>,
+        RE: Error + 'static,
+    {
+        let details = self.device_authorization_details.ok_or_else(|| {
+            RequestTokenError::Other("no device authorization details provided".to_string())
+        })?;
+
+        let mut elapsed = Duration::new(0, 0);
+        let mut interval = details.interval().clone();
+
+        loop {
+            if elapsed > details.expires_in() {
+                return Err(RequestTokenError::Other("Device code expired".to_string()));
+            }
+
+            // Prepare the request.
+            match http_client(self.prepare_request()?)
+                .await
+                .map_err(|_| DeviceCodeAction::DoubleIntervalThenRetry)
+                .and_then(device_token_action)
+            {
+                Ok(http_response) => break token_response(http_response),
+                Err(DeviceCodeAction::Retry) => {
+                    // Wait for the current interval.
+                }
+                Err(DeviceCodeAction::IncreaseIntervalThenRetry) => {
+                    // Increase the interval, then wait that long.
+                    interval = interval + Duration::from_secs(5);
+                }
+                Err(DeviceCodeAction::DoubleIntervalThenRetry) => {
+                    // Double the interval, then wait that long.
+                    interval = interval.checked_mul(2).ok_or(RequestTokenError::Other(
+                        "Failed to increase interval".to_string(),
+                    ))?;
+                }
+                Err(DeviceCodeAction::NoFurtherRequests(req)) => break token_response(req),
+            };
+
+            // Use async-std to sleep asynchronously.
+            async_std::task::sleep(interval).await;
+            elapsed = elapsed + interval;
```

Comment:
`elapsed` isn't quite accurate here because it doesn't account for the time spent making the HTTP requests and waiting for the server to respond. it also doesn't account for any time that elapses between the client receiving the device authorization response and calling `request`/`request_async` to enter the polling loop.

in most cases, the server will tell us when the device code has expired by returning `expired_token`. the pathological case we have to worry about is where the HTTP request repeatedly fails (e.g., device goes offline, the server returned a bad URL, etc.). in that case, we'll continue to loop until the device token expiration is reached, which could be a very long time if the server returns a long `expires_in`. the current approach allows a bad server to trivially DoS the client, which is especially problematic for the sync version of this interface.

to avoid these issues, I suggest two changes:

1. add a `timeout: Option<Duration>` argument to both `request` and `request_async`. this allows callers to specify an upper-bound for the duration of the polling loop
2. compute `elapsed` on each loop iteration by comparing the current time against the start time of the polling loop. since time needs to be mockable, I suggest we use `chrono::Utc::now()` by default for this purpose but add a `set_time_fn` method to `DeviceAccessTokenRequest` so that it can be overridden. For an example of this pattern, see https://github.com/ramosbugs/openidconnect-rs/blob/a86baefa8fcdda975a74d115ce6c3e5a2a12d721/src/verification.rs#L660-L671

Reply:
Fixed this up. Haven't added a test yet though, will do.